File: bun-hono/src/middleware/auth.middleware.ts
import { Context, Next } from 'hono'
import JwtUtils from '../utils/jwt.utils'
import AppError from '../common/api.error'

export default class AuthMiddleware {
  static async authenticate(c: Context, next: Next) {

    const authHeader = c.req.header('Authorization')
    if (!authHeader?.startsWith('Bearer ')) {
      throw new AppError(401, 'Unauthorized')
    }
    const token = authHeader.split(' ')[1]
    try {
      const payload = await JwtUtils.verifyToken(token)
      if (!payload) {
        throw new AppError(401, 'Unauthorized')
      }
      c.set('userId', payload.id)
      c.set('email', payload.email)
      await next()
    } catch (err) {
      throw new AppError(401, 'Unauthorized')
    }
  }
}

File: go-gin/pkg/middleware/auth_middleware.go
package middleware

import (
	"strings"

	"github.com/dist-r/rcontacts-rest/go-gin/pkg/app"
	"github.com/dist-r/rcontacts-rest/go-gin/pkg/utils"
	"github.com/gin-gonic/gin"
)

func AuthMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		// Get token from Authorization header
		authHeader := c.GetHeader("Authorization")
		if authHeader == "" {
			c.Error(app.NewError(401, "Unauthorized"))
			return
		}

		// Extract token (assuming "Bearer <token>")
		token := authHeader[len("Bearer "):]
		if token == "" {
			c.Error(app.NewError(401, "Invalid authorization header"))
			return
		}

		// Check if token starts with "Bearer "
		if !strings.HasPrefix(token, "Bearer ") {
			c.Error(app.NewError(401, "Invalid authorization header"))
			return
		}

		// Remove "Bearer " prefix
		token = strings.TrimPrefix(token, "Bearer ")

		// Verify token and get userId from claims
		claims, err := utils.VerifyToken(token)
		if err != nil {
			c.Error(err)
			return
		}

		// Set user ID in context
		c.Set("user_id", claims.UserID)

		c.Next()
	}
}
